src/textual/widgets/_progress_bar.py

```diff
@@ -0,0 +1,432 @@
+"""Implements a progress bar widget."""
+
+from __future__ import annotations
+
+from functools import partial
+from math import ceil
+from time import monotonic
+from typing import Optional
+
+from rich.style import Style
+
+from ..app import ComposeResult, RenderResult
+from ..containers import Horizontal
+from ..message import Message
+from ..reactive import reactive
+from ..renderables.bar import RenderableBar
+from ..timer import Timer
+from ..widget import Widget
+from ..widgets import Label
+
+
+class Bar(Widget, can_focus=False):
+    """The bar portion of the progress bar."""
+
+    COMPONENT_CLASSES = {"bar--bar", "bar--complete", "bar--indeterminate"}
+    """
+    | Class | Description |
+    | :- | :- |
+    | `bar--bar` | Style of the bar (may be used to change the color). |
+    | `bar--complete` | Style of the bar when it's complete. |
+    | `bar--indeterminate` | Style of the bar when it's in an indeterminate state. |
+    """
+
+    DEFAULT_CSS = """
+    Bar {
+        width: 32;
+        height: 1;
+    }
+    Bar > .bar--bar {
+        color: $warning;
+        background: $foreground 10%;
+    }
+    Bar > .bar--indeterminate {
+        color: $error;
+        background: $foreground 10%;
+    }
+    Bar > .bar--complete {
+        color: $success;
+        background: $foreground 10%;
+    }
+    """
+
+    _completion_percentage: reactive[float | None] = reactive[Optional[float]](None)
+    """The percentage of the bar that should be filled or `None` if indeterminate.
+
+    If the completion percentage is `None`, then the bar will render as
+    an indeterminate progress bar.
+    """
+    _refresh_timer: Timer | None
```

Comment:
This should really be initialized in the constructor, but there doesn't seem to be one.

I think we need a `__init__` for this reason, and that we want to prevent the dev from passing in child widgets.

Reply:
Sure thing.